### PR TITLE
Calibration assistant

### PR DESCRIPTION
**Pull request by bwdev01**

### CODE REVIEW

File: myframe_events.cpp
Comment (agalasso) [inline myframe_events.cpp:1047]: should we also say "run the Calibration Assistant to improve results" here too? the people who hit this error are the ones who especially need to know!

File: myframe_events.cpp
Comment (agalasso) [inline myframe_events.cpp:1044]:
```suggestion
                    else if (dec > degrees(Scope::DEC_COMP_LIMIT))
```

File: myframe_events.cpp
Comment (agalasso) [inline myframe_events.cpp:1032]: should we move this block to a helper function? Then this code becomes:
```
    bool proceed = CheckForCalibration();
    if (proceed)
        StartGuiding();
```
```
static bool
CheckForCalibration()
{
    if (TheScope()->IsCalibrated())
        return true;
    double ra, dec, lst;
    if (!pPointingSource->GetCoordinates(&ra, &dec, &lst))
    {
        ...
```

File: phdcontrol.cpp
Comment (agalasso) [inline phdcontrol.cpp:68]: can we invert this to be a positive attribute to reduce double negatives?
```suggestion
    bool useStickyLock;
```

File: phdcontrol.cpp
Comment (agalasso) [inline phdcontrol.cpp:120]: let's make this a positive attribute too (and if we converge recalibrate and useStickyLock into an `options` param with option bits, then the option bit could be positive like USE_STICKY_LOCK_POS)
```suggestion
bool PhdController::Guide(bool recalibrate, const SettleParams& settle, const wxRect& roi, bool useStickyLockPos, wxString *error)
```

File: scope.cpp
Comment (agalasso) [inline scope.cpp:1006]: since it is not obvious, could we add a comment here explaining what this means?

**Review by agalasso — COMMENTED** [review]
good stuff Bruce!

File: calibration_assistant.cpp
Comment (agalasso) [inline calibration_assistant.cpp:355]:
```suggestion
```

File: calibration_assistant.cpp
Comment (agalasso) [inline calibration_assistant.cpp:294]: optional: you could eliminate much of the indentation to improve readability by adding some early returns:
```
    if (!pPointingSource || !pPointingSource->IsConnected())
        return;

    double raSpd, decSpd;
    bool error = pPointingSource->GetGuideRates(&raSpd, &decSpd);
    if (error)
        return;

    if (!pPointingSource->ValidGuideRates(raSpd, decSpd))
        return;

    ... and so forth ...

File: calibration_assistant.cpp
Comment (agalasso) [inline calibration_assistant.cpp:419]: I'd also decrease indentations here in the same way (I find deep indentation difficult to read since I lose the context of the branches)
```
    bool error = pPointingSource->GetCoordinates(&ra, &dec, &lst);
    if (error)
    {
        // show error message
        return;
    }
    // ... ordinary non-error code follows ...
    ...
    if (m_meridianFlipping)
        return;
    ... ordinary non-meridian flipping stuff ...
    etc
}

File: calibration_assistant.cpp
Comment (agalasso) [inline calibration_assistant.cpp:478]:
```suggestion
        if (!pPointingSource || !pPointingSource->CanReportPosition())
            throw ERROR_INFO("CalPositionRecommendations: mount not connected or not reporting pointing info");
```

File: calibration_assistant.cpp
Comment (agalasso) [inline calibration_assistant.cpp:487]: C/C++ can do nasty things with implicit conversions so I find it helpful to use explicitly-typed constants rather than letting the compiler do implicit casting. 
In this case that means
```suggestion
            if (hourAngle <= 0.)
```
likewise for 30 => 30. 90 => 90. below    (30. and 30.0 would both be ok -- both are the same double whereas 30 is int)

File: calibration_assistant.cpp
Comment (agalasso) [inline calibration_assistant.cpp:530]: should we maybe tone down "as close as possible" -- I'm imagining people obsessively trying to zero-out their dec unnecessarily.
Maybe something like: "Use the Slew button to move the scope to within 15 degrees or so of Dec = 0"

File: calibration_assistant.cpp
Comment (agalasso) [inline calibration_assistant.cpp:537]: I guess this is a pre-existing message but I think this message is confusing for newbs who won't know what dec comp is.  Maybe we should be more direct and just tell them to calibrate closer to dec 0, or to disable dec comp?

File: calibration_assistant.cpp
Comment (agalasso) [inline calibration_assistant.cpp:533]:
```suggestion
        m_pExplanation->SetLabelText(_("Calibration is likely to fail this close to the pole.\n") + slewCond);
```
This comment is for all the translatable messages here: let's not include layout formatting ("\n" or trailing " ") in the message that is being translated -- translators may or may not overlook that when they translate and we want the code to be in control of the layout.
Let's ensure that the translators are presented with sentences they can translate without regard to layout formatting. That way they can focus on the translation proper , and the code can do the layout.

File: calibration_assistant.cpp
Comment (agalasso) [inline calibration_assistant.cpp:547]: ditto comment about deep indentation vs early returns

**Review by agalasso — COMMENTED** [review]
thanks for making those changes.   I had a chance to look at calibraton_assistant.cpp and have some mostly cosmetic suggestions there.

File: eegg.cpp
Comment (bwdev01) [reply eegg.cpp:243]: This is a poor man's "singleton" solution much like the GA and other non-modal dialogs.  We really only want one instance of the CA active.  The destructor for CA nulls the member variable in Frame so I think the null check works as we want.

File: calibration_assistant.cpp
Comment (bwdev01) [reply calibration_assistant.cpp]: This is funny.  I intended it to mean "as close as practical for your site" but as you say, we would have people spending 20 minutes trying to completely zero out the Dec.  We know that many people have significant site visibility problems, some quite severe.  Hopefully the changes will work for both cases.

File: calibration_assistant.cpp
Comment (bwdev01) [reply calibration_assistant.cpp]: I've taken a different approach using the 'Wrap' method, much like the profile wizard.  Beyond your points, there's really no way to know how things will appear after translation.  So the long text messages are now wrapped and embedded line feeds have been removed.

File: eegg.cpp
Comment (agalasso) [reply eegg.cpp:243]: @bwdev01  I get the singleton thing, that's all good.
The observation is that by line 242 pCalibrationAssistant is never null, thus we can remove the check for null on line 242.

File: calibration_assistant.cpp
Comment (agalasso) [inline calibration_assistant.cpp:1019]: should we ask for this value in hours then convert to degrees internally?  probably most planetarium and imaging apps will report hour angle in hours not degrees?
we could even ask for the target RA and compute hour angle ourselves, right? (or maybe not, that's assuming they want to image the target _now_)

Comment (bwdev01) [reply calibration_assistant.cpp:1019]: A lot of this is for users who don't understand the sky coordinate system and have no concept of why right ascension and declination are in different units.  Someone who is dealing with sky obstructions can probably have an intuitive notion of x degrees from the "center" but they aren't going to translate that into x/15 hours.  I don't think a planetarium app has any meaning here, it's PHD2 that's going to slew the scope.  We're not going to point at a particular target, we're going to point "over there".  In my view, changing the hour angle will be an unusual case and mostly done to deal with obstruction problems (balcony location, trees, neighbor's house...).

File: calibration_assistant.cpp
Comment (bwdev01) [reply calibration_assistant.cpp:875]: That was the idea of the 'explain' button.  If they want to know how to get a better grade, they would click on 'explain' and get info on what the specific underlying problem is.  I think the alert levels we have provide the necessary warnings and are important enough that we can defend raising the concern.  What I see on the forum is that any sort of caution or warning causes users to become paralyzed, yell for help or complain on various other forums, perhaps only to be told that they should just forge ahead.  I think dealing in generalities here would in fact open up a can of worms.  "Acceptable" means just that - "it's ok to proceed".

File: calibration_assistant.cpp
Comment (bwdev01) [reply calibration_assistant.cpp:135]: I'm not following this one.  These variables are members of the CalCustomDialog class and they appear at the top of that class definition.  What am I missing?

File: calibration_assistant.cpp
Comment (bwdev01) [reply calibration_assistant.cpp:869]: All I wanted here was a 'thumbs-up' message, didn't want to get into TMI.  People don't like calibration.  Setting auto-restore really depends on whether they set up every night and are being careful to reassemble the guiding assembly the same way each night.  Even the "any sky location" behavior depends on having Dec comp set - admittedly nearly always true.  I can be convinced if you feel strongly about it but sometimes less said is more.

File: calibration_assistant.cpp
Comment (agalasso) [reply calibration_assistant.cpp:135]: I was suggesting putting the data members before method declarations, i.e. move lines 130-135 (the data member decls) to after line 125

File: calibration_assistant.cpp
Comment (agalasso) [reply calibration_assistant.cpp:869]: agreed, don't want to fall int a TMI trap, which my suggested message probably does.
I am imagining the newbie person running the calibration wiz and getting this message, and then wondering "now what?".   Was trying to suggest we tell them what to do next. Any better ideas?

File: calibration_assistant.cpp
Comment (agalasso) [reply calibration_assistant.cpp:875]: hmm, could we actually say "It's ok to proceed."?  that could be helpful! (assuming they know what proceed means -- they might not)

File: calibration_assistant.cpp
Comment (agalasso) [reply calibration_assistant.cpp:1019]: ok.    I think what threw me off was "target" -- when I saw that I thought "imaging target" -- but this is really the calibration target location.  Should we avoid the word target to avoid mixing it up with imaging target?  Maybe call it "Calibration position" or "Calibration location" or "Calibration pointing position"?

File: calibration_assistant.cpp
Comment (bwdev01) [reply calibration_assistant.cpp:135]: Ok, I see it now - I couldn't see they were below the constructor declaration.

File: calibration_assistant.cpp
Comment (bwdev01) [reply calibration_assistant.cpp:869]: At this point, guiding has started.  I suppose I could say guiding has started using the new calibration.  I didn't want to close the dialog because the user might want to iterate or experiment further and I wanted them to see a ta-da message.

File: calibration_assistant.cpp
Comment (bwdev01) [reply calibration_assistant.cpp:1019]: Ok, yes, "target" is probably a loaded term.  I will try "Calibration location".  This may also help discourage the notion that you need to choosing a specific star for calibration.

File: calibration_assistant.cpp
Comment (agalasso) [inline calibration_assistant.cpp:288]: nit
```suggestion
    *PrefHA = pConfig->Profile.GetInt("/scope/CalSlew/TgtHA", defBestOffset);
```

File: calibration_assistant.cpp
Comment (agalasso) [inline calibration_assistant.cpp:341]: minor
```suggestion
    if (!msg.empty())
```

File: calibration_assistant.cpp
Comment (agalasso) [inline calibration_assistant.cpp:459]: nit
```suggestion
}

```

File: calibration_assistant.cpp
Comment (agalasso) [inline calibration_assistant.cpp:785]: optional:
```suggestion
    if (vec.empty())
```

File: calibration_assistant.cpp
Comment (agalasso) [inline calibration_assistant.cpp:1066]: This needs to be a single translatable string, not two separate strings -- the individual pieces may not make any sense translated then concatenated together.
The translator is going to use poedit and get a list of all the strings marked by `_()` and translate them one by one.
The two fragments make sense in English concatenated together, but that does not necessarily mean it will work in other languages. For example, It may be more idiomatic in some other language to phrase it like "specify here the sky position required by your location conditional" but as you have it here it forces the translator to translate the individual pieces ("if your site...", "you can spcify") then concatenates them together which prevents the translator from coming up with the best wording for the full sentence.
```suggestion
    pMessage->SetLabelText(_("If your site location requires a unique sky position for calibration, "
                             "you can specify it here."));
```

File: calibration_assistant.cpp
Comment (agalasso) [inline calibration_assistant.cpp:1156]: nit
```suggestion
```

File: calibration_assistant.cpp
Comment (agalasso) [inline calibration_assistant.cpp:1182]: nit
```suggestion
    pHeader->SetLabelText(_("Assuming you have followed all the recommendations of the Calibration Assistant, "
```

File: myframe_events.cpp
Comment (agalasso) [reply myframe_events.cpp:1068]: @bwdev01 I think this suggestion is still applicable - the code can be simplified by removing the extraneous check for null on line 1067

**Review by agalasso — COMMENTED** [review]
thanks for addressing my comments!
Only a couple things I believe we should address before merge
 - the translation issue about the concatenated strings
 - making the tooltips imperative to be consistent with other tooltips in the app

the remaining comments are minor and should not block merging this, but feel free to consider the suggestions

**Review by agalasso — APPROVED** [review]
perfect, thanks!

**Comment by bwdev01** [comment]
Thanks for the help.

**Pull request merged**
